seqr/utils/vcf_utils.py

```diff
@@ -0,0 +1,97 @@
+import re
+
+from collections import defaultdict
+
+from seqr.utils.middleware import ErrorsWarningsException
+from seqr.utils.file_utils import file_iter
+
+BLOCK_SIZE = 65536
+
+EXPECTED_META_FIELDS ={
+    'INFO': {
+        'AC': {
+            'Number': 'A',
+            'Type': 'Integer',
+        },
+        'AN': {
+            'Number': '1',
+            'Type': 'Integer',
+        },
+        'AF': {
+            'Number': 'A',
+            'Type': 'Float',
+        }
+    },
+    'FORMAT': {
+        'GQ': {
+            'Number': '1',
+            'Type': 'Integer',
+        },
+        'GT': {
+            'Number': '1',
+            'Type': 'String',
+        }
+    }
+}
+
+REQUIRED_HEADERS = ['#CHROM', 'POS', 'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO', 'FORMAT']
+
+
+def _validate_vcf(header, meta, first_line):
+
+    missing_headers = [h for h in REQUIRED_HEADERS if h not in header ]
+    if missing_headers:
+        missing_fields = ','.join(missing_headers)
+        raise ErrorsWarningsException([f'Required VCF header fields {missing_fields} are missing.'], [])
+
+    for field, sub_field_meta in EXPECTED_META_FIELDS.items():
+        for sub_field, metadata in sub_field_meta.items():
+            for key, value in metadata.items():
+                if meta.get(field, {}).get(sub_field, {}).get(key) != value:
+                    error = f'VCF header field {field}.{sub_field} and meta information {key}={value} is expected.'
+                    raise ErrorsWarningsException([error], [])
+
+    row_data = first_line.split('\t')
+    info_keys = [k.split('=')[0] for k in row_data[7].split(';')]
+    missing_info_fields = [f for f in EXPECTED_META_FIELDS['INFO'].keys() if f not in info_keys ]
```

Comment:
I'm a bit confused by whats happening here. You seem to be validating that INFO and FORMAT exist by checking both the `meta` and the `first_line` for the exact same values. Per the discussion in the ticket, it would be better if we use the meta to do validation, if we can check everything in the meta we should not check the `first_line` at all

Reply:
Even if the subfields exist in the `meta`, they are still possibly missing in the data lines. So it is safer to validate both. But since each data line can be different, it is not a complete validation to check the `first_line`. I'm okay with taking out the validation for the first line.